# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/WebrtcVideoCodecFactory.h
# Commit: c34ea4d71235
# Full Hash: c34ea4d71235c0519020893ba04d7206e7fe4f14
# Author: Michael Froman <mfroman@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 1
# Description:
#   Bug 1654112 - fix tsan error, using a destroyed mutex. r=ng
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D121007
# ==============================================================================

diff -r f547e3673690 -r c34ea4d71235 dom/media/webrtc/libwebrtcglue/WebrtcVideoCodecFactory.h
--- a/dom/media/webrtc/libwebrtcglue/WebrtcVideoCodecFactory.h	Wed Jul 21 07:51:58 2021 -0700
+++ b/dom/media/webrtc/libwebrtcglue/WebrtcVideoCodecFactory.h	Mon Jul 26 16:42:40 2021 -0500
@@ -15,33 +15,35 @@
  public:
   GmpPluginNotifier()
       : mOwningThread(GetCurrentSerialEventTarget()),
-        mCreatedGmpPluginEvent(mOwningThread),
-        mReleasedGmpPluginEvent(mOwningThread) {}
+        mCreatedGmpPluginEvent(
+            new MediaEventForwarder<uint64_t>(mOwningThread)),
+        mReleasedGmpPluginEvent(
+            new MediaEventForwarder<uint64_t>(mOwningThread)) {}
 
   ~GmpPluginNotifier() {
     mOwningThread->Dispatch(NS_NewRunnableFunction(
         "~GmpPluginNotifier",
         [createdEvent = std::move(mCreatedGmpPluginEvent),
          releasedEvent = std::move(mReleasedGmpPluginEvent)]() mutable {
-          createdEvent.DisconnectAll();
-          releasedEvent.DisconnectAll();
+          createdEvent->DisconnectAll();
+          releasedEvent->DisconnectAll();
         }));
   }
 
   MediaEventSource<uint64_t>& CreatedGmpPluginEvent() {
     MOZ_ASSERT(mOwningThread->IsOnCurrentThread());
-    return mCreatedGmpPluginEvent;
+    return *mCreatedGmpPluginEvent;
   }
 
   MediaEventSource<uint64_t>& ReleasedGmpPluginEvent() {
     MOZ_ASSERT(mOwningThread->IsOnCurrentThread());
-    return mReleasedGmpPluginEvent;
+    return *mReleasedGmpPluginEvent;
   }
 
  protected:
   const nsCOMPtr<nsISerialEventTarget> mOwningThread;
-  MediaEventForwarder<uint64_t> mCreatedGmpPluginEvent;
-  MediaEventForwarder<uint64_t> mReleasedGmpPluginEvent;
+  RefPtr<MediaEventForwarder<uint64_t> > mCreatedGmpPluginEvent;
+  RefPtr<MediaEventForwarder<uint64_t> > mReleasedGmpPluginEvent;
 };
 
 class WebrtcVideoDecoderFactory : public GmpPluginNotifier,
@@ -87,8 +89,9 @@
  public:
   explicit WebrtcVideoEncoderFactory(std::string aPCHandle)
       : mInternalFactory(MakeUnique<InternalFactory>(std::move(aPCHandle))) {
-    mCreatedGmpPluginEvent.Forward(mInternalFactory->CreatedGmpPluginEvent());
-    mReleasedGmpPluginEvent.Forward(mInternalFactory->ReleasedGmpPluginEvent());
+    mCreatedGmpPluginEvent->Forward(mInternalFactory->CreatedGmpPluginEvent());
+    mReleasedGmpPluginEvent->Forward(
+        mInternalFactory->ReleasedGmpPluginEvent());
   }
 
   std::vector<webrtc::SdpVideoFormat> GetSupportedFormats() const override {
